DAOS-9088 docker: add recipes to run in containers (#7556)
daos-stack/daos: utils/docker/vcluster/README.md

- The `daos-admin` node allowing to manage the DAOS server thanks to `dmg`command.
- The `daos-client` node using the the DAOS server to store data.

!!! note
    At this time only emulated hardware storage are supported by this Docker platform:
    - SCM (i.e. Stoarage Class Memory) are emulated with standard RAM memory.
    - NVMe disks are emulated with a file device.

!!! warning
    Virtual Docker network such as [bridge](https://docs.docker.com/network/bridge/) are not yet
    well supported by DAOS.  Thus, one physical network interface of the host should be choosen for
    being used by the containers through the Docker
    [host](https://docs.docker.com/network/host/) network.

## Prerequisites

To build and deploy the Docker images, `docker` and optionally `docker-compose` shall be available.
The docker host should have access to the [Docker Hub](https://hub.docker.com/) and
[CentOS](https://www.centos.org/) offical repositories.  Finally,
[hugepages](https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt) linux kernel feature shall
be enabled on the docker host.

The platform was tested and validated with the [CentOS8](https://hub.docker.com/_/centos) official

> docker image.  However other RHEL like distributions such as

Reviewer: s/RHEL like/RHEL-like/

Author: fixed with commit https://github.com/daos-stack/daos/pull/7556/commits/445b4543d7d9ee5f98b5e6de5eca826f0c160b37

[rocky8.4](https://hub.docker.com/r/rockylinux/rockylinux) should be supported.

!!! warning
    Some distributions are not yet well supported such as
    [rocky8.5](https://hub.docker.com/r/rockylinux/rockylinux): issue with the management of
    hugepages with the [spdk](https://spdk.io/) library.

## Building Docker Images

### Building Base DAOS Image

The first image to create is the `daos-base` image which is not intetended to be used as it, but as
a base image for building the other three daos images.  This first image could be built directly
from GitHub with the following command:

```bash
$ docker build --tag daos-base:centos8 \
	"https://github.com/daos-stack/daos.git#master:utils/docker/vcluster/daos-base/el8"
```

or from a local tree with the following command:

```bash
$ docker build --tag daos-base:centos8 "utils/docker/vcluster/daos-base/el8"
```

> This Docker file accept the following arguments:

Reviewer: Maybe explain how the arguments can be passed on a cmd line by giving an example.

Author: fixed with commit https://github.com/daos-stack/daos/pull/7556/commits/445b4543d7d9ee5f98b5e6de5eca826f0c160b37

Add paragraph containing an example showing how to enable authentication.

- `DAOS_BASE_IMAGE`: Base docker image to use (default daos-base)
- `DAOS_BASE_VERSION`: Version of the base docker image to use (default centos8)
- `DAOS_AUTH`: Enable DAOS authentification when set to "yes" (default "no")
- `DAOS_HUGEPAGES_NBR`: Number of huge pages to allocate for SPDK (default 4096)
- `DAOS_IFACE_NAME`: Fabric network interface used by the DAOS engine (default "eth0")
- `DAOS_SCM_SIZE`: Size in GB of the RAM emulating SCM devices (default 4)
- `DAOS_BDEV_SIZE`: Size in GB of the file created to emulate NVMe devices (derault 16)

The Dockerfile of the `daos-client` and `daos-admin` images accept the following arguments:
- `DAOS_BASE_IMAGE`: Base docker image to use (default daos-base)
- `DAOS_BASE_VERSION`: Version of the base docker image to use (default centos8)
- `DAOS_AUTH`: Enable DAOS authentification when set to "yes" (default "no")
- `DAOS_ADMIN_USER`: Name or uid of the daos administrattor user (default "root")
- `DAOS_ADMIN_GROUP`: Name or gid of the daos administrattor group (default "root")

From a local, tree they could also be built with the following commands:

```bash
$ cd "utils/docker/vcluster"
$ docker-compose build

> ```

Reviewer: Could you please also provide the command to build from the GitHub directly?

Author: fixed with commit fixed with commit https://github.com/daos-stack/daos/pull/7556/commits/445b4543d7d9ee5f98b5e6de5eca826f0c160b37

Add paragraph containing an example showing how to build these images directly from GitHub.

    images and base image).

## Running DAOS Nodes Container

Once the images are created, the containers could be directly started with docker with the following
commands:

```bash
$ export DAOS_IFACE_IP=x.x.x.x
$ docker run --detach --privileged --name=daos-server --hostname=daos-server \
	--add-host "daos-server:$DAOS_IFACE_IP" --add-host "daos-admin:$DAOS_IFACE_IP" \
	--add-host "daos-client:$DAOS_IFACE_IP" --volume=/sys/fs/cgroup:/sys/fs/cgroup:ro \
	--volume=/dev/hugepages:/dev/hugepages  --tmpfs=/run --network=host \
	daos-server:centos8
$ docker run --detach --privileged --name=daos-agent --hostname=daos-agent \
	--add-host "daos-server:$DAOS_IFACE_IP" --add-host "daos-admin:$DAOS_IFACE_IP" \
	--add-host "daos-client:$DAOS_IFACE_IP" --volume=/sys/fs/cgroup:/sys/fs/cgroup:ro \
	--tmpfs=/run --network=host daos-agent:centos8
$ docker run --detach --privileged --name=daos-client --hostname=daos-client \
	--add-host "daos-server:$DAOS_IFACE_IP" --add-host "daos-admin:$DAOS_IFACE_IP" \
	--add-host "daos-client:$DAOS_IFACE_IP" --volume=/sys/fs/cgroup:/sys/fs/cgroup:ro \
	--tmpfs=/run --network=host daos-client:centos8

> ```

Reviewer: That's very long commands ... shall we also provide a simple script that setup the containers, format DAOS, create a pool, create a container and mount it in the client container? I assume that this is what most folks might want to try first.

Author: Fixed with commit fixed with commit https://github.com/daos-stack/daos/pull/7556/commits/445b4543d7d9ee5f98b5e6de5eca826f0c160b37

Add bash script _daos-cm.sh_ allowing to easily manage the virtual cluster.
Add a section describing how to use this bash script.

To build and deploy the Docker images, `docker` and optionally `docker-compose` shall be available.
The docker host should have access to the [Docker Hub](https://hub.docker.com/) and
[CentOS](https://www.centos.org/) offical repositories.  Finally,
[hugepages](https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt) linux kernel feature shall
be enabled on the docker host.

The platform was tested and validated with the [CentOS8](https://hub.docker.com/_/centos) official
docker image.  However other RHEL-like distributions such as
[rocky8.4](https://hub.docker.com/r/rockylinux/rockylinux) should be supported.

!!! warning
    Some distributions are not yet well supported such as
    [rocky8.5](https://hub.docker.com/r/rockylinux/rockylinux): issue with the management of
    hugepages with the [spdk](https://spdk.io/) library.

## Building Docker Images

### Building Base DAOS Image

The first image to create is the `daos-base` image which is not intetended to be used as it, but as
a base image for building the other three daos images.  This first image could be built directly
from GitHub with the following command:

```bash
$ docker build --tag daos-base:centos8 \

> 	"https://github.com/daos-stack/daos.git#master:utils/docker/vcluster/daos-base/el8"

Reviewer 2: Have you tried this?  I get a submodule error when I run it here.

Author: I tested it with replacing the uri of my branch.
This last one will only work when the PR will be merged.

This should be replaced with
docker build --tag daos-base:centos8 \
	"https://github.com/knard-intel/daos.git#DEV/DAOS-9088:utils/docker/vcluster/daos-base/el8"

Reviewer 2: I was using docker build "https://github.com/knard-intel/daos.git#DEV/DAOS-9088:utils/docker/vcluster/daos-base/el8"

and the error I get is

unable to prepare context: unable to 'git clone' to temporary context directory: error initializing submodules: usage: git submodule [--quiet] add [-b <branch>] [-f|--force] [--name <name>] [--reference <repository>] [--] <repository> [<path>]
   or: git submodule [--quiet] status [--cached] [--recursive] [--] [<path>...]
   or: git submodule [--quiet] init [--] [<path>...]
   or: git submodule [--quiet] deinit [-f|--force] [--] <path>...
   or: git submodule [--quiet] update [--init] [--remote] [-N|--no-fetch] [-f|--force] [--rebase] [--reference <repository>] [--merge] [--recursive] [--] [<path>...]
   or: git submodule [--quiet] summary [--cached|--files] [--summary-limit <n>] [commit] [--] [<path>...]
   or: git submodule [--quiet] foreach [--recursive] <command>
   or: git submodule [--quiet] sync [--recursive] [--] [<path>...]
: exit status 1

Author: Do not understand, I tries from my personal computer at home and it is working.
Which version of docker and git do you have ?

From my side , I tried with the following version of docker: Docker version 20.10.11, build dea9396
And the following version of git: git version 2.25.1 and git version 2.27.0

```

or from a local tree with the following command:

```bash

> $ docker build --tag daos-base:centos8 "utils/docker/vcluster/daos-base/el8"

Reviewer 2: Do you need quotes here?
```suggestion
$ docker build --tag daos-base:centos8 utils/docker/vcluster/daos-base/el8
```

Author: Not really, habit from me to always quote file path.
If needed, I could remove all these useless quote.

Author: fixed

[CentOS](https://www.centos.org/) official repositories.  Finally,
[hugepages](https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt) linux kernel feature shall
be enabled on the docker host.

The platform was tested and validated with the [CentOS8](https://hub.docker.com/_/centos) official
docker image.  However other RHEL-like distributions such as
[rocky8.4](https://hub.docker.com/r/rockylinux/rockylinux) should be supported.

!!! warning
    Some distributions are not yet well supported such as
    [rocky8.5](https://hub.docker.com/r/rockylinux/rockylinux): issue with the management of
    hugepages with the [spdk](https://spdk.io/) library.

## Building Docker Images

### Building Base DAOS Image

The first image to create is the `daos-base` image which is not intetended to be used as it, but as
a base image for building the other three daos images.  This first image could be built directly
from GitHub with the following command:

```bash
$ docker build --tag daos-base:centos8 \
	https://github.com/daos-stack/daos.git#master:utils/docker/vcluster/daos-base/el8
```

or from a local tree with the following command:

```bash

> $ docker build --tag daos-base:centos8 utils/docker/vcluster/daos-base/el8

Reviewer 2: This command should be run from the root of the git checkout, which is itself fine but is inconsistent with the below.

Reviewer 2: ```suggestion
$ cd utils/docker/cluster
$ docker build --tag daos-base:centos8 daos-base/el8
```

Author: Fixed with commit https://github.com/daos-stack/daos/pull/7556/commits/8ccbf36ca07b92b342baba440f407062e54862fb
Remove all cd command: all command are self contained :)

- `DAOS_AUTH`: Enable DAOS authentication when set to "yes" (default "no")
- `DAOS_REPOS`: Space separated list of repos needed to install DAOS (default
	"https://packages.daos.io/v2.0")
- `DAOS_GPG_KEYS`: Space separated list of GPG keys associated with DAOS repos (default
	"https://packages.daos.io/RPM-GPG-KEY")
- `DAOS_REPOS_NOAUTH`: Space separated list of repos to use without GPG authentication
	(default "")

For example, building a DAOS base image with authentication enabled could be done with the
following command:

```bash
$ docker build --tag daos-base:centos8 --build-arg DAOS_AUTH=yes \
	https://github.com/daos-stack/daos.git#master:utils/docker/vcluster/daos-base/el8
```

### Building DAOS Nodes Images

The three images `daos-server`, `daos-admin` and `daos-client` could be built directly from GitHub
or from a local tree in the same way as for the `daos-base` image.  Following command could be used
to build directly the three images from GitHub:

```bash
$ for image in daos-server daos-admin daos-client ; do \
	docker build --tag "$image:centos8" \
		"https://github.com/daos-stack/daos.git#master:utils/docker/vcluster/$image/el8"; \
  done

> ```

Reviewer 2: I wondered if this step was necessary, but it may be if there isn't a way for the docker-compose command to be run from github, I found it slightly confusing to have an optional step in the middle of the instructions.

Author: Indeed, docker-compose does not yet support github as input :'(
From what I understand of the US, it was required that all step should be able to be run without having access to the source tree.

The Docker file of the `daos-server` image accept the following arguments:

- `DAOS_BASE_IMAGE`: Base docker image to use (default daos-base)
- `DAOS_BASE_VERSION`: Version of the base docker image to use (default centos8)
- `DAOS_AUTH`: Enable DAOS authentication when set to "yes" (default "no")
- `DAOS_HUGEPAGES_NBR`: Number of huge pages to allocate for SPDK (default 4096)

> - `DAOS_IFACE_NAME`: Fabric network interface used by the DAOS engine (default "eth0")
> - `DAOS_SCM_SIZE`: Size in GB of the RAM emulating SCM devices (default 4)
> - `DAOS_BDEV_SIZE`: Size in GB of the file created to emulate NVMe devices (derault 16)

Reviewer 2: It would be handy here to have a note that the IP of this interface will be moved later, perhaps re-order these so you can have more text for DAOS_IFACE_NAME without losing the visual fact that these are a block.

Minor, but "lo" might be preferably here to "eth0".

You also have a spelling mistake here.  I'm not verified the marking rending of this proposed change.
```suggestion
- `DAOS_SCM_SIZE`: Size in GB of the RAM emulating SCM devices (default 4)
- `DAOS_BDEV_SIZE`: Size in GB of the file created to emulate NVMe devices (default 16)
- `DAOS_IFACE_NAME`: Fabric network interface used by the DAOS engine (default "eth0")
  The IP address for this interface will be required when starting DAOS.
```

Author: I have some issue with the loopback interface at the beginning.
I am going to test it again: indeed it will be more portable.

Author: Yes, I have some issue with hwloc and the loopback interface (as for the docker bridge network).
I am going to add a statement on this point in the limitation warning.
@johannlombardi is aware of this limitation with hwloc. It is planned to fix it later. Please @johannlombardi , could you confirm this last point.

Author: > It would be handy here to have a note that the IP of this interface will be moved later, perhaps re-order these so you can have more text for DAOS_IFACE_NAME without losing the visual fact that these are a block.
> 
> Minor, but "lo" might be preferably here to "eth0".
> 
> You also have a spelling mistake here. I'm not verified the marking rending of this proposed change.

Going to integrate your remark for the interface. However, do not find the spelling mistake.

Reviewer 2: "derault"

Author: Arghhh, thanks

Author: fixed with commits:
- https://github.com/daos-stack/daos/pull/7556/commits/83667f55edd452cc0e00bfd93c2952c46ef4a951
- https://github.com/daos-stack/daos/pull/7556/commits/f1b52574e79de789a736911b401f0b55e424915f

Author: > Sadly, I have some issue with hwloc and the loopback interface (as for the docker bridge network). I am going to add a statement on this point in the limitation warning. @johannlombardi is aware of this limitation with hwloc. It is planned to fix it later. Please @johannlombardi , could you confirm this last point.

Add comment on loopback support issue in commit https://github.com/daos-stack/daos/pull/7556/commits/9972f9cab1b697b03b5ebda59a6d89ee03f90c7b

The Dockerfile of the `daos-client` and `daos-admin` images accept the following arguments:

- `DAOS_BASE_IMAGE`: Base docker image to use (default daos-base)
- `DAOS_BASE_VERSION`: Version of the base docker image to use (default centos8)
- `DAOS_AUTH`: Enable DAOS authentication when set to "yes" (default "no")
- `DAOS_ADMIN_USER`: Name or uid of the daos administrattor user (default "root")
- `DAOS_ADMIN_GROUP`: Name or gid of the daos administrattor group (default "root")

> From a local tree, a more straightforward way to build these images could be done with

Reviewer 2: Should this be on L94?  Currently you have the following sections here:

1. How to build from github
2. What options are accepted
3. How to build from a checkout

Would it be clearer to have 1, 3, then 2?

Author: Indeed, it is better like this: more over this is the layout of the next section.

Author: Going to change it.

Author: Fixed with commit https://github.com/daos-stack/daos/commit/83667f55edd452cc0e00bfd93c2952c46ef4a951

`docker-compose` and the following commands:

```bash
$ cd utils/docker/vcluster

> $ docker-compose build

Reviewer 2: This command requires being run from a directory which is not the root of the git checkout, which is inconsistent with the above.

Author: Fixed: see previous comment.

	--add-host "daos-client:$DAOS_IFACE_IP" --volume=/sys/fs/cgroup:/sys/fs/cgroup:ro \
	--tmpfs=/run --network=host daos-client:centos8
```

The value of the `DAOS_IFACE_IP` shall be replaced with the one of the network interface which was
provided when the images have been built.

Once started, the DAOS server waits for the administrator to format the system.
This can be done using the following command:

```bash
$ docker exec daos-admin dmg -i storage format
```

Upon successful completion of the format, the storage engine is started, and pools
can be created using the daos admin tool.  For more advanced configurations and usage refer to the
section [DAOS Tour](https://docs.daos.io/QSG/tour/).

### Starting Containers with docker-compose command

From a local tree, a more straightforward way to start the containers could be done with
`docker-compose` and the following commands:

```bash
$ cd utils/docker/vcluster
$ docker-compose up --detach
```

!!! note
    Before starting the containers with `docker-compose`, the IP address of the network interface,
    which was provided when the images have been built, shall be defined in the Docker

>     Compose environment file `utils/docker/vcluster/.env`.

Reviewer 2: I seem to have missed this step, yet I was able to run fio anyway?

Author: Not needed if you use the daos-cm helper script.
Only needed, if you want to use directly docker-compose.

As with the docker command, the system shall be formatted, pools created, etc..

### Managing Virtual Docker cluster with daos-cm.sh

From a local tree, the bash script `utils/docker/daos-cm.sh` could be used to start the containers
and setup a simple DAOS system composed of the following elements:

- 1 DAOS pool of 10GB (i.e. size of the pool is configurable)
- 1 DAOS POSIX container mounted on /mnt/daos-posix-fs

This script could also be used to respectively stop and monitor the containers.

More details on the usage of `daos-cm.sh` command could be found with running the following command:

```bash

> $ bash utils/docker/daos-cm.sh --help

Reviewer 2: This path is wrong, and can we just set execute permissions on the file?  Following the above examples...
```suggestion
$ cd utils/docker/vcluster
$ ./daos-cm.sh --help
```

Author: It is working from anywhere: there si a cd just before the switch-case.
https://github.com/knard-intel/daos/blob/8ccbf36ca07b92b342baba440f407062e54862fb/utils/docker/vcluster/daos-cm.sh#L224

Reviewer 2: It does work from anywhere, but the script will not work if you call it as utils/docker/daos-cm.sh as it's not in the utils/docker directory.

Author: My bad, I did not understood your remark.
fixed with commit https://github.com/daos-stack/daos/pull/7556/commits/61862ea6fb4e6a26720ca90f8a71ee4ea8c1ad32
